### ISSUE TITLE
False positive: Unnecessary type casting

### ISSUE THREAD

**Issue by voku** [kalessil/phpinspectionsea#1672]
| Subject        | Details                                                                       |
| :------------- | :---------------------------------------------------------------------------- |
| Plugin         | Php Inspections (EA Ultimate), 3.0.10 |
| Language level | PHP 7.2                 |

Current behaviour
--
```php
    /**
     * @return array<int|string, string>
     */
    function foo(): array {
      // ...
    }
    foreach (foo() as $key => value) {
       if (strpos((string)$key, "foo") !== false) {
           echo "foo";       
       }
    }
```

I see this ```[EA] This type casting is not necessary, as the argument is of needed type.``` but the code is 100% valid and the string cast is needed.

Expected behaviour
--
no report about the type casting

Environment details
---
PhpStorm 2020.3
Build #PS-203.5981.175, built on December 2, 2020
Licensed to Portable UTF-8 / Lars Moelleken
Subscription is active until February 27, 2021
For non-commercial open source development only.
Runtime version: 11.0.9+11-b1145.21 amd64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
Linux 4.15.0-128-generic
GC: G1 Young Generation, G1 Old Generation
Memory: 5120M
Cores: 4
Registry: run.processes.with.pty=TRUE, ide.tooltip.initialDelay=885, ide.balloon.shadow.size=0
Non-Bundled Plugins: DBN, Key Promoter X, String Manipulation, awesome.console, com.alayouni.ansiHighlight, com.andrey4623.rainbowcsv, com.intellij.ideolog, dev.arunvelsriram.desccron, net.seesharpsoft.intellij.plugins.csv, uk.co.ben-gibson.remote.repository.mapper, com.jetbrains.codeWithMe, name.kropp.intellij.makefile, pronskiy.elephpant, NEON support, com.kalessil.phpStorm.phpInspectionsUltimate, me.artspb.idea.eval.plugin, net.king2500.plugins.PhpAdvancedAutoComplete, lv.midiana.misc.phpstorm-plugins.deep-keys, de.espend.idea.php.toolbox, de.espend.idea.php.annotation, fr.adrienbrault.idea.symfony2plugin, izhangzhihao.rainbow.brackets, ru.adelf.idea.dotenv
Current Desktop: ubuntu:GNOME

**Comment by ea-inspections-team** [comment]
Thanks for reporting @voku!
Probably it was caused by types incorrectly resolved by IDE.  The issue is not reproducible with the latest IDE/analyzer.
